Paper: UDL: Unified Interface for Deep Learning (Elmarakeby)

marakeby opened the pull request:
Scipy 2015 paper submission
UDL: Unified Interface for Deep Learning
Haitham Elmarakeby and Lenwood S. Heath

[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/28

michaelpacer commented:
I could be wrong, but I think all paper merge requests are supposed to begin with `Paper: `.

marakeby commented:
Thanks 
I cannot find this in the submission guidelines. Anyway, I changed the title.

ankurankan commented:
@marakeby A few suggestions:

1. Please add references for the libraries in the style as in https://raw.githubusercontent.com/scipy-conference/scipy_proceedings/2015/papers/00_vanderwalt/00_vanderwalt.rst.

2. In the introduction you have mentioned about 'designing a network that has five consecutive layers'. It would be difficult for someone who doesn't have a background in deep learning / neural network to understand what you are referring to. Maybe having a short introduction to deep learning first would be a nice addition.

3. You have talked about configuration files in the Introduction but haven't mentioned anywhere which configuration files are you talking about.

4. Maybe restructuring the paper a bit might help the readers understand things more clearly. Moving parts of caffe and pylearn sections before discussing the details of UDL will help the readers understand the short comings in using these libraries and what problem UDL is trying to solve more clearly.

Please let me know your thoughts on these points.

marakeby commented:
Thanks for the comments. These are valid points. I am working on fixing them

amueller commented:
I think lasagna, keras and nolearn are all going in the direction of defining architectures in code instead of configuration files. I am not really an expert, but from what I understand, these are widely-used libraries, so you should at least mention them, and probably compare to them.

amueller commented on papers/Haitham_Elmarakeby/Haitham_Elmarakeby.rst:
```diff
@@ -0,0 +1,181 @@
+The classes layer simply encapsulates meta-classes, exposes the attributes using Python properties, and provides initialization and default values.
+Although this layer may go out of synchronization if the base Caffe library changes, we depend on the community to keep it up to date.
+
+Layers act as low level building blocks, so we combined these layers into higher level models that can do certain job.
```
do *a* certain job.

marakeby replied:
thanks, corrected

amueller commented on papers/Haitham_Elmarakeby/Haitham_Elmarakeby.rst:
```diff
@@ -0,0 +1,181 @@
+If the child implementation does not set these two attributes, it has to override all the *UDLModel* functions.
+Each library implementation is organized under different directories; models, data sets, and trainers. The models directory contains all the core classes of the library.
+All classes in this directory inherit from *UDLModel* except for helper classes that are used to define the core models such as layers in the case of Caffe.
+The UDL model class encapsulates the functionality of the underlying library class. For example the *Autoencoder* class in UDL.pylearni corresponds to the *Auoencoder* class in Pylearn2.
```
typo in second instance of Autoencoder.

marakeby replied:
Thanks, corrected

amueller commented on papers/Haitham_Elmarakeby/Haitham_Elmarakeby.rst:
```diff
@@ -0,0 +1,181 @@
+UDL models encapsulate the corresponding Caffe messages and provide an easy way for initialization and accessing their attributes.
+
+
+Many model classes in the underlying libraries do not know how to train themselves. However, in UDL model classes, it is required to be able to train yourself.
```
I don't understand this. PyLearn2 and caffe  come with a set of learning algorithms, right?

marakeby replied:
Sure, Pylearn2 and Caffe come with training algorithms. The point is that the core models such as Autoencoder, RBM, DBN, MLP, not all of them can train themselves. A trainer or solver has to be invoked to train this model. For example Pylearn2 provides a speical script to train models. In Scikit-learn things are diffenent. The model has to be able to train itself e.g. in the fit function. Scikit-learn require the model to be self contianed and this what we did in UDL. I hope this clarifies the point little bit, please feel free to ask again if i faild to make it clear.

amueller commented on papers/Haitham_Elmarakeby/Haitham_Elmarakeby.rst:
```diff
@@ -0,0 +1,181 @@
+--------
+Pylearn2 is a deep learning library based on Theano [1]_ that can compile models into C++ code that can run on GPUs or CPUs.
+In Pylearn2, a model can be defined in a YAML markup language or simply in Python.
+Pylearn2 provides an excellent python interface but it does not officially support defining models in code.
```
This sentence seems to be contradictory with the previous. I think it is not really possible to configure a network "in code" without writing YAML, right?

marakeby replied:
Pylearn2 provide python interface, you can define a model without wrting YAML  but they does not officially support this kind of development. The provided interface is not compatible with Scikit-learn. It is written in the Pylearn2 vision document the following, "One goal we used to have but no one seems to actually work is supporting the scikit-learn interface. It could make sense to add support for the the scikit-learn interface to one of our models such as S3C, RBMs, or autoencoders"
https://github.com/lisa-lab/pylearn2/blob/d43d015d11747bdbd49737ce1163b9af67eac840/doc/vision.txt

jaberg commented:
@marakeby sorry we didn't get reviews in earlier, but @amueller and I will get you some feedback today.

jaberg commented:
(thanks @amueller! for helping to wrap things up!)

amueller commented on papers/Haitham_Elmarakeby/Haitham_Elmarakeby.rst:
```diff
@@ -0,0 +1,181 @@
+
+The extended model keeps the same signature of the parent model, which means that it can be easily used the same way the parent is used.
+Any extra initialization arguments are added as optional parameters. In particular two extra parameters are needed to enable the model to train itself, *Trainer* and *dataset_adapter*.
+The base *UDLModel* has a *get_param* function; however, the Pylearn2 model defines a function with the same name.
```
I'm not sure what you mean by changing the name of the function. On the UDL class? Or are you monkey-patching pylearn2?
I am not sure how relevant this is for the end user. The goal of UDL is to provide a unified interface so that users don't have to deal with the caffe or pylearn2 interface, right?

marakeby replied:
I think you are correct, I feel it should not be mentioned in the paper. I removed the sentence. Anyway what happens is that there was a conflict between the Pylearn2 and Scikit-learn, both of them define a function with the same name for different purposes. As a work around we changed the function name of Pylearn2 in our local copy. This workaround can be pushed to the end user as a small script that renames all the occurrences of this function inside the code repository.

amueller replied:
On 07/01/2015 07:22 PM, marakeby wrote:
>
> In papers/Haitham_Elmarakeby/Haitham_Elmarakeby.rst 
> <https://github.com/scipy-conference/scipy_proceedings/pull/128#discussion_r33735309>:
>
> I think you are correct, I feel it should not be mentioned in the 
> paper. I removed the sentence. Anyway what happens is that there was a 
> conflict between the Pylearn2 and Scikit-learn, both of them define a 
> function with the same name for different purposes. As a work around 
> we changed the function name of Pylearn2 in our local copy. This 
> workaround can be pushed to the end user as a small script that 
> renames all the occurrences of this function inside the code repository.
>
>
I still don't see why you need to do this.
In your class you can be explicit about which parent class you want to 
call a function on, or you can use super carefully.

amueller commented on papers/Haitham_Elmarakeby/Haitham_Elmarakeby.rst:
```diff
@@ -0,0 +1,181 @@
+    from sklearn.pipeline import Pipeline
+    from sklearn.metrics import accuracy_score
+
+    def test(X, Y, Yt):
```
I wouldn't necessarily put this in a function, because it doesn't really add additional information. Also, Xt is not defined. Maybe use ``X_train, X_test, y_train, y_test``

marakeby replied:
Thanks, corrected

amueller commented:
@jaberg Well I missed 2 of my reviews and 1 was late, that is the least I can do ;)

amueller commented:
I feel there is a central part missing: how do you specify complex models in UDL?
You only gave an example of a simple auto-encoder and and a logistic regression.

Both pylearn2 and caffe allow the construction of complex convolutional or fully connected nets. How does the YAML translate to python code? How do you specify lenet? How a densely connected network with different setting of dropout or maxout etc. ?

Is there a layer abstraction as in torch, keras and theano?

As you goal is providing a common interface that hopefully captures many usecases, I feel you should focus much more on what this interface is. Here, by interface I don't mean the ``fit``, ``predict`` API of scikit-learn, but how the architecture and training method can be expressed in UDL.

amueller commented on papers/Haitham_Elmarakeby/Haitham_Elmarakeby.rst:
```diff
@@ -0,0 +1,205 @@
+Caffe, on the other hand, provides a set of Python messages that cannot be inherited.
+UDL models encapsulate the corresponding Caffe messages and provide an easy way for initialization and accessing their attributes.
+
+Many classes in the underlying libraries do not know how to train themselves. However, UDL model need to be able to train itself.
```
Well, maybe I would phrase it differently then. The division of concerns between Pylearn2 and scikit-learn is just different. That can be handled pretty easily by instantiating a solver in ``fit``, though, right? How do you specify the solver in UDL? And is that uniform between the two backends?

marakeby replied:
UDL is wrapping the solvers of caffe and pylearn2. The solver is passed as a parameters to the model in the constructor. we have different solvers for different libraries.

amueller commented on papers/Haitham_Elmarakeby/Haitham_Elmarakeby.rst:
```diff
@@ -0,0 +1,205 @@
+--------
+Pylearn2 is a deep learning library based on Theano [Ber10]_ that can compile models into C++ code that can run on GPUs or CPUs.
+In Pylearn2, a model can be defined in a YAML markup language or simply in Python.
+Pylearn2 provides an excellent python interface but it does not officially support defining models in code.
```
I'm not sure what "officially" means. Does it implement this or not? Not being scikit-learn compatible is different than not having a python interface. Lasagne has a python interface but is not scikit-learn compatible, nolearn wraps lasagne with a scikit-learn compatible interface.

marakeby replied:
Pylearn2 has a python interface that is not compatible with Scikit-learn.

amueller replied:
Exactly. That is not what the paper says.

amueller commented:
I'm not sure if the paper in the current form warrants inclusion in the scipy proceedings. The title promises a common interface, an I was expecting a way to generically define networks. There is nothing in the paper about how to define networks, though.

What UDL delivers is a scikit-learn compatible wrapper of pylearn2 and caffe.
I haven't found online documentation and the classes don't seem to have docstrings, though.

marakeby commented:
It seems that there is some miscommunication here. The goal UDL is not to define a new syntax for developers to use instead of other libraries. The goal of UDL is to let developers define their models in pure python using their preferred library and allow these models to interact with each other in memory with ease.

marakeby commented:
here is an example of how to write LeNet using the caffe python interface
import caffe
from udl.caffei.datasets.DummyDataset import DummyDataset
from udl.caffei.datasets.DummyDataset_batch import DummyDataset_batch
from udl.caffei.datasets.MemoryDataset import MemoryDataset
from udl.caffei.models.layers.innerproduct_pb2 import InnerProduct
from udl.caffei.models.layers.layer import Phase, Layer
from udl.caffei.models.layers.loss_layers import Loss, LossType
from udl.caffei.models.layers.network import Network
from udl.caffei.models.utils import get4D
from udl.caffei.trainers.SGDTrainer import SGDTrainer
from udl.model import UDLModel
import udl.caffei.models.caffe_pb2 as pb2
from udl.caffei.models.layers.hidden_layers import Convolution, Pooling, Relu, LRN, InnerProduct, Filler
import numpy as np

class LeNet(UDLModel):
    def __init__(self,  dataset_adaptor=DummyDataset_batch(), trainer=SGDTrainer()):
        caffe.set_mode_cpu()
        self.dataset_adaptor = dataset_adaptor
        self.trainer = trainer


    def fit(self, X, Y):
        X4D = get4D(X)
        Y4D = get4D(Y)
        data_layer = self.dataset_adaptor.fit_transform(X4D,Y4D)
        net = self.get_net(data_layer)
        self.solver = self.trainer.get_trainer(net,X4D, Y4D)
        self.trainer.solve()
        # self.solver.set_train_data (X4D, Y4D)
        # self.solver.solve()
        return self.solver
        

    def predict(self, X):
        X4D = get4D(X)
        self.solver.set_test_data (X4D)
        self.solver.test_nets[0].forward()
        prob = self.solver.test_nets[0].blobs['output'].data
        print prob.shape
        # pred = prob[:, 0] < prob[:, 1]
        pred= np.argmax(prob, axis=1)
        return pred

    def get_net(self, data_layer):
        net = Network()

        # data_layer =  MemoryDataset().fit_transform()

        net.add(data_layer)
        net.name = "LeNet"
        c_bottom = ['data', 'pool0']
        c_num_output = [20, 50]

        max= pb2.PoolingParameter.MAX
        xavier = pb2.FillerParameter(type= 'xavier')
        const =  pb2.FillerParameter(type= 'constant')
        for i in range(0,2):
            c = 'conv%d'%i
            p = "pool%d"%i
            r = "relu%d"%i
            n = "norm%d"%i
            c_layer = Convolution(name=c , bottom= c_bottom[i] , num_output = c_num_output[i] , kernel_size =5, stride=1, weight_filler= xavier , bias_filler = const)
            p_layer = Pooling(name= p, bottom= c, kernel_size=2 , stride =2, pool = max )
            net.add(c_layer), net.add(p_layer),

        ip1 = InnerProduct(name="ip1", bottom="pool1", num_output=500, weight_filler= xavier , bias_filler = const)
        r_layer = Relu(name = r, bottom= "ip1",  top= "ip1")
        ip2 = InnerProduct(name="ip2", bottom="ip1", num_output=10,  weight_filler= xavier , bias_filler = const)

        o1 = Loss(name="loss", type=LossType.SoftmaxWithLoss, bottom=["ip2", "label"], include=Phase.TRAIN)
        o2 = Loss(name="output", type= LossType.SOFTMAX, top= "output", bottom="ip2", include=Phase.TEST)

        net.add(ip1), net.add(r_layer), net.add(ip2), net.add(o1), net.add(o2)


        return net.str().__str__()

amueller commented:
Then I clearly misunderstood the point. I would not call this unified, though.
If you want to use the same lenet with the pylearn2 interface, you have to rewrite it from scratch.

> The goal UDL is not to define a new syntax for developers to use instead of other libraries. The goal of UDL is to let developers define their models in pure python using their preferred library and allow these models to interact with each other in memory with ease.

What do you mean by interacting in memory with ease? Supporting the scikit-learn interface?

marakeby commented:
yes

marakeby commented:
@ankurankan Thanks for your constructive comments
The paper should be ready now. I have fixed the references. I have added an introduction to deep learning. I pushed more examples with real data to the examples directory (https://github.com/marakeby/udl/tree/master/udl/examples). I have added a figure to explain how repeated pattern in network can captures in code better than text-based configuration file. I use 'configuration file' to mean the file in which developer define his network (model). Let me know if you have more comment.

shjb commented:
This paper looks almost ready to go, the comments of the reviewers seem to have been addressed.
Only thing missing is citation for Pylearn2. (@marakeby please add this)

amueller commented:
I feel the paper should make more clear what the purpose of UDL is. It doesn't seem like a very unified interface to me.

shjb commented:
@marakeby -- I've looked into this paper more carefully, and I'd like to update my assessment from 4 hours ago that this paper is ready to go.  The paper is certainly improved, but when scipy papers are about new packages, the packages themselves are part of the submission and the review process. Some of @amueller's comments are about the code and the design of UDL rather than the paper, and these have not been addressed.

Looking at https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md
I think @amueller is challenging whether this work, including both the paper and the UDL library has sufficient novelty and impact:

"The work should employ new or innovative methods or approaches to a problem.
The work should advance the state of a scientific domain, the practice of scientific computing itself, or another subject area within the scope of the conference."

This is a judgement call of course, but I tend to agree with @amueller .  The purpose of a unified interface is to abstract away the details that relate to the libraries you use under the hood. In contrast, the code sample above about lenet includes many references to `caffei`.  If a search and replace for `s/caffei/pylearn2i` isn't sufficient to switch implementations, then you haven't finished making a unified interface.

Meanwhile, the authors of python bindings for caffe and the authors of pylearn have worked hard to make their respective libraries easy to use, modular, etc. What's possible with UDL that wasn't before? For me, the most compelling use-case for UDL is the pipeline setup that you describe, which combines Caffe, Pylearn and Sklearn. However, sklearn's pipeline construct does not (afaik) support training by back-propagation, the benefit of combining modules this way is significantly undermined.

I don't think I am bringing up new points that were not already raised by @amueller. I'm simply re-stating @amueller's points in a way that might help clear up why they have arguably *not* been addressed by the updates you've made to the paper.

shjb commented:
Hello @marakeby,

Proceedings co-chairs @katyhuff and I have discussed this paper, and we agree with the reviewers that this paper and library are not ready for inclusion in the scipy proceedings.

Personally, I'm a huge fan of deep learning software though, and I hope you continue to work on this promising project. I hope you will consider re-submitting this work to scipy as it matures and gathers traction in the community.
